Improve CI

Roms1383 opened the pull request:
Related to #5

Roms1383 commented:
I have a suggestion regarding the `publish crate` job in `main` `release` workflow: usually you might want to run everything on every commit (`check` / `build` / `format` / `lint`) and only run `release` related job steps on `main` once all the other jobs have succeeded. Would you like me to make the appropriate changes in that regard ?

Odonno commented:
Hi @Roms1383 

Thank you for tackling this issue. With your help, I want to improve the whole CI, improve quality/reliability and reduce build time as much as possible.

Indeed, I want the `publish crate` to run once everything succeed. That's why it's the last command in the job. I suppose you want to make a separate/dependent job? If so, feel free to do so.

I already have some question regarding the current state of your PR:

* [ ] Should the cargo config file be named `config.toml`? (I see the extension is missing)
* [ ] Is `git-fetch-with-cli = true` supposed to fix the `error: failed to verify the checksum of surrealdb v1.0.0-beta.9+20230402` error?
* [ ] I hope it is not too much to ask but `format` and `lint` are currently not parts of the CI. Is it possible for you to add them?
* [ ] Same question for `check` step if you think that's necessary to add it

Anyway, thanks again for doing this.

Roms1383 commented:
> Should the cargo config file be named config.toml? (I see the extension is missing)

cargo actually reads both (see [config](https://doc.rust-lang.org/cargo/reference/config.html)), but you're right the preferred  one is with `.toml` so I updated it.

> Is git-fetch-with-cli = true supposed to fix the error: failed to verify the checksum of surrealdb v1.0.0-beta.9+20230402 error?

Yes, as far as I remember there's an issue with semver in dependencies version at the moment. I fixed it on my repo by temporarily specifying this and the git path (see e1becfd).

> I hope it is not too much to ask but format and lint are currently not parts of the CI. Is it possible for you to add them?

Of course! :)

> Same question for check step if you think that's necessary to add it

Added it too!

Roms1383 commented:
By the way you don't necessarily need to use char emoji directly in commit messages, but you can type them as text like`:emoji:` instead ? See [gitmoji](https://gitmoji.dev/).

Roms1383 commented:
How do you usually like to trigger releases ?
By committing a tag on `main` branch, or via Github releases ?

Odonno commented:
> By the way you don't necessarily need to use char emoji directly in commit messages, but you can type them as text like`:emoji:` instead : see [gitmoji](https://gitmoji.dev/).

I am a lazy guy, I click on the icon to copy/paste the emoji.

Odonno commented:
> How do you usually like to trigger releases ? By committing a tag on `main` branch, or via Github releases ?

My current process is the following:

* Change the version in `Cargo.toml` via a commit, and publish a new crate version only if a new version is detected on `main` branch (main workflow)
* Manually create a GitHub release following the pushed commit to `main`

I am comfortable to have a manual process for the moment. I could change it if you have provide an interesting solution.

Roms1383 commented:
Mine is similar:
- bump version in `Cargo.toml` on `main`, then commit + push
- add `tag` on `main` then commit + push and let workflows do the rest (build > bundle artifacts > create draft github release)

and if everything goes fine I manually update the draft release to a published release.

For example in this [release workflow](https://github.com/Roms1383/mdbook-codeblocks/blob/main/.github/workflows/release.yml).

Odonno commented:
So, you would move the publish crate step to the release workflow? If this is your idea, I am ok doing that.

Roms1383 commented:
Should be good to review :)

Roms1383 commented on Cargo.toml:
```diff
@@ -30,7 +30,7 @@ serde_json = "1.0"
 tokio = { version = "1.26.0", features = ["full"] }
 sqlparser = "0.33.0"
-surrealdb = "1.0.0-beta.9"
+surrealdb = { git = "https://github.com/surrealdb/surrealdb.git", tag = "v1.0.0-beta.9+20230402" }
```
this is temporary until https://github.com/rust-lang/cargo/issues/7180 is fixed or until surrealdb bumps to next non-semver version.

Odonno replied:
Can you target the `main` branch so we can fix #20 ?

Roms1383 replied:
why would you prefer targeting `main` branch ? usually it's better to target `tag`, if I understand correctly.

Odonno replied:
There is currently an important fix that lives in the `main`. Without that, the lib is pretty useless in production. I know this is a bad practice but it is temporary until the beta 10 version.

Roms1383 replied:
Ok I've done the change locally: pinning on `main` is failing the tests. I'm currently investigating.

Roms1383 replied:
Do you have an idea why `tarpaulin` is failing because of missing files / folders ?

Please look at this [action run on my fork](https://github.com/Roms1383/surrealdb-migrations/actions/runs/5052684992)
This is the action run for:
- commit ffc2654
- prior to bumping `surrealdb` to `main`
- with `tarpaulin` enabled from `pull_request`).

Roms1383 replied:
Otherwise typical error I get in failed tests when switching on `main` is:

failures:

```sh
---- cli::apply::up::apply_initial_migrations stdout ----
Error: Unexpected failure.
code-1
stderr=
Error: There was a problem with the database: Parse error on line 5 at character 28 when parsing \'TYPE record<user> ASSERT $value != NONE;
DEFINE FIELD created_at ON post TYPE datetime VALUE $before\'

Caused by:
    There was a problem with the database: Parse error on line 5 at character 28 when parsing \'TYPE record<user> ASSERT $value != NONE;
    DEFINE FIELD created_at ON post TYPE datetime VALUE $before\'
```
Still looking out why ^^

Roms1383 replied:
Ok waiting for ur feedback on both.

I couldn't figure out which file is actually missing in `tarpaulin` coverage, and it's especially weirder since it works just fine on my local.

On the other hand, I've tried an extensive bunch of stuff to find out why when switching to `surrealdb main` there are parse errors, till now unsuccessfully.

Odonno replied:
Ok, I looked at the code of `surrealdb` and record type now looks like this `record<user>` or this `record<type1 | type2 | type3>`. I beleive they are using the discriminated union notation from TypeScript. Not sure what to do here, either

* stick to the current beta-9 version with the existing bug
* or move to a `0.10.X` release right now by changing the templates

I think the best move would be to rollback to beta-9 for now, merge the PR, release the v0.9.6 and then make changes to release preview for `0.10.X`.

Roms1383 replied:
I was wondering exactly the same, and I even tried to change `record(T)` for `record<T>` everywhere but it didn't fix (except if I missed some autogenerated part, that could be).

And concerning `tarpaulin`, any idea why it complains over and over that `Error: No such file or directory` ?

Odonno replied:
Regarding the missing files/folder. I have now idea. If the code is strictly similar to `main` branch, it should work. The CI tests job is not 100% reliable, you can try to re-run the job and see if you have a different result.

Roms1383 replied:
Tests now pass on `main`, currently investigating about `tarpaulin`.

Odonno replied:
Yes, I know the CI for tests is very long. I hope we can find ways to reduce the build time in the future.

Roms1383 commented on .cargo/config.toml:
```diff
@@ -0,0 +1,2 @@
+[net]
+git-fetch-with-cli = true
```
this is temporary until https://github.com/rust-lang/cargo/issues/7180 is fixed or until surrealdb bumps to next non-semver version.

Roms1383 commented on src/scaffold/schema.rs:
```diff
@@ -79,7 +79,7 @@ fn scaffold_from_schema(
 
     let schema = convert_ast_to_surrealdb_schema(ast, preserve_casing)?;
 
-    if schema.tables.len() <= 0 {
```
here `.len()` is a `usize` so it cannot be under `0`, then `clippy` suggests `is_empty` for readability.

Odonno replied:
I should use `isEmpty()` instead of `len()` whenever I can. That's on me. Thanks for the fix.

Roms1383 reviewed (commented):
Ok I gave a couple more context.

Odonno commented on .github/workflows/main.yml:
```diff
@@ -1,51 +1,55 @@
-  push:
-    branches:
-      - "main"
+on: push
```
Why remove the constraint on "main" branch? The `main` workflow is for when pushing to `main`. The `pull-request` worklow if for when we push to a PR branch.

If the 2 workflows are the same, we can remove the `pull-request` and have a single `main` workflow.

Roms1383 replied:
Yes actually I came to the same conclusion, that `pull-request` workflow was redundant, let me remove it.

Roms1383 replied:
done in 7ff2aba

Odonno commented on .github/workflows/main.yml:
```diff
@@ -1,51 +1,55 @@
   coverage:
+    needs: ci
     runs-on: ubuntu-latest
+    if: github.ref == 'refs/heads/main'
```
Should not be here, per the last review comment.

Roms1383 replied:
well this one is because you might not want coverage on each and every push to each and every branch, hence why I initially restricted it to tags on `main`.

Odonno replied:
Branch coverage is the detection of every paths in a codebase. Like when you encounters an `if`. Without branch coverage, having a single test that goes in the if, whether the condition is `true` or `false`, will be 100% valid coverage. This is a nice feature to detect the different cases but I can work without it for the moment.

However diff/patch is extremely useful because it allows you to see if merging a branch will increase or decrease the overall coverage of the codebase, like you can see on other PR.

Odonno commented on src/config/core.rs:
```diff
@@ -1,5 +1,6 @@
 use super::common::retrieve_config_value;
 
+#[allow(dead_code)]
```
I assume the CI fails if there is dead code? I need to refactor the code and make an underlying lib project that can be used by either `main.rs` or `lib.rs`. If it fails, I'm ok to leave it. Anyway, this is a temporary solution.

Roms1383 replied:
Yes indeed there's a bunch of dead or duplicated functions (e.g. `create_surrealdb_client` which appears both in lib and tests alike). Putting annotations allow clippy to succeed, and are good reminders that code needs to be cleaned.

Odonno reviewed (changes requested):
Good job on this. Also, when you are done with all the changes, can you rebase your commit history to clean it?

Odonno commented:
I suppose there are many changes in the code due to `cargo check`? I am using the `rust-lang.rust-analyzer` vscode extension. Do you recommend a better extension and/or is it compatible with `cargo check`?

Roms1383 commented:
> I suppose there are many changes in the code due to cargo check? I am using the rust-lang.rust-analyzer vscode extension. Do you recommend a better extension and/or is it compatible with cargo check?

That's the recommended one, I use it too, it should be fine.

Roms1383 commented on .github/workflows/release.yml:
```diff
@@ -33,3 +39,9 @@ jobs:
           ARCHIVE_TYPES: ${{ matrix.archive }}
           EXTRA_FILES: "readme.md LICENSE"
+
+      - name: Publish crate
```
done in 22fe4b9

Roms1383 replied:
let me know if I should remove the restriction on `tag` too.

Roms1383 commented:
> Good job on this. Also, when you are done with all the changes, can you rebase your commit history to clean it?

done in 63e8d45 and 535fa89

Roms1383 commented:
Ok please hold on a bit @Odonno, currently rebasing and then I'll have to address your reviews :)

By the way I have a question: is there a reason to duplicate methods `create_surrealdb_client`, `create_surrealdb_connection`, `sign_in` and `set_namespace_and_database` in `surrealdb.rs` files both `src` and `tests` ? They look identical at first sight.

Odonno commented:
They are pretty much the same functions. Indeed, some code is duplicated between `src` and `tests`. I want to improve the architecture with a dedicated library. I will create a new issue for this issue. Anyway, I don't think this change should fit in this PR.

Roms1383 commented:
> They are pretty much the same functions. Indeed, some code is duplicated between src and tests. I want to improve the architecture with a dedicated library. I will create a new issue for this issue. Anyway, I don't think this change should fit in this PR.

Sounds good to me 👌

Roms1383 commented:
Ready for review @Odonno 🎉 
CI is running [on fork over there](https://github.com/Roms1383/surrealdb-migrations/actions/runs/5058495735).

Roms1383 commented:
As a side note, a recurrent warning in the console inform that `license-file` and `license` keys are redundant in `Cargo.toml` : whichever would you like to keep ?

Odonno commented:
> As a side note, a recurrent warning in the console inform that `license-file` and `license` keys are redundant in `Cargo.toml` : whichever would you like to keep ?

Yes, I know. I started to play with `license-file` but instead of showing `MIT` in the crates.io website, it showed `unknown` or something like that. I added `license = MIT` to force the display of MIT on this website. Not sure what is the best option.

Roms1383 commented:
You could just keep `license = MIT` without deleting the `LICENSE` file at the root of the repo, maybe ?
This way `crates.io` shows up correctly and user still have access to your `LICENSE` file.
or just look at how other popular crates out there handle that :)
It's out of the scope of this PR anyway.

As you can see from [last 2 Actions runs](https://github.com/Roms1383/surrealdb-migrations/actions), we're green & ready for review ! 🚀

Odonno commented on tests/library/validate_version_order.rs:
```diff
@@ -87,7 +87,9 @@ async fn fails_if_migrations_applied_with_new_migration_before_last_applied() ->
 
             let first_migration_file = get_first_migration_file()?;
-            std::fs::remove_file(first_migration_file)?;
+            if first_migration_file.exists() {
```
I suppose the `if` should not be here. If the `remove_file` function fails, the test should fail if we can't find the file.

Roms1383 replied:
I can revert that. I wrongfully thought the `No such file or directory` error was related, while it was due to missing SurrealDB in workflow.

Roms1383 replied:
reverted in 400619f

Odonno commented on tests/helpers/io.rs:
```diff
@@ -23,7 +23,7 @@ pub fn empty_folder(folder: &str) -> Result<()> {
     let dir = Path::new(folder);
 
-    if dir.exists() {
+    if dir.is_dir() && dir.exists() {
```
Return an error if `dir` is not a dir?

Roms1383 replied:
Same as my previous comment: I can revert that. I wrongfully thought the No such file or directory error was related, while it was due to missing SurrealDB in workflow.

Roms1383 replied:
reverted in e01c565

Odonno commented on tests/helpers/io.rs:
```diff
@@ -12,7 +12,7 @@ pub fn clear_tests_files() -> Result<()> {
     let migrations_files_dir = Path::new(folder);
 
-    if migrations_files_dir.exists() {
+    if migrations_files_dir.is_dir() && migrations_files_dir.exists() {
```
Return an error if `migrations_files_dir` is not a dir?

Roms1383 replied:
Same as my previous comment: I can revert that. I wrongfully thought the No such file or directory error was related, while it was due to missing SurrealDB in workflow.

Roms1383 replied:
reverted in e01c565

Odonno commented on tests/helpers/io.rs:
```diff
@@ -33,7 +33,7 @@ pub fn remove_folder(folder: &str) -> Result<()> {
     let schemas_files_dir = Path::new("tests-files/schemas");
 
-    if schemas_files_dir.exists() {
+    if schemas_files_dir.is_dir() && schemas_files_dir.exists() {
```
Return an error if `schemas_files_dir` is not a dir?

Roms1383 replied:
Same as my previous comment: I can revert that. I wrongfully thought the No such file or directory error was related, while it was due to missing SurrealDB in workflow.

Roms1383 replied:
reverted in e01c565

Odonno commented:
> You could just keep `license = MIT` without deleting the `LICENSE` file at the root of the repo, maybe ? This way `crates.io` shows up correctly and user still have access to your `LICENSE` file. or just look at how other popular crates out there handle that :) It's out of the scope of this PR anyway.
> 
> As you can see from [last 2 Actions runs](https://github.com/Roms1383/surrealdb-migrations/actions), we're green & ready for review ! 🚀

Review is done. There are many changes in this PR so I detected some things to change.

I will remove `license-file` property then.

Roms1383 commented on .github/workflows/main.yml:
```diff
@@ -1,57 +1,70 @@
   coverage:
+    needs: ci
     runs-on: ubuntu-latest
+    if: github.ref == 'refs/heads/main' || github.event_name != 'pull_request'
```
all or?

Roms1383 replied:
removed in 4b466ea

Roms1383 commented on .github/workflows/main.yml:
```diff
@@ -1,57 +1,70 @@
         with:
           version: "0.22.0"
-          args: "--engine llvm"
+          args: "--engine llvm --target-dir target/tarpaulin"
```
on subsequent builds (after the first one) it's making `coverage` faster (30min is  way too long)

Roms1383 replied:
I linked related articles in the commit, if you want to read them out.

Roms1383 replied:
I added `--skip-clean` as suggested in the related articles, as `--target-dir` alone this does not seem to speed up the build (see running actions duration on my fork).

Roms1383 replied:
in my own book, waiting 30min for code coverage is just unacceptable 😄

Roms1383 replied:
as it turns out from latest runs, `--target-dir` / `--skip-clean` does not seem to speed up the build anytime 😒 ...

Odonno replied:
https://identeco.de/en/blog/speed-up-rust-ci-pipelines-that-use-tarpaulin/

I understand now. So setting `target-dir` allows separate build artifacts between a normal build and a test build.

It seems that the cache is not saved between those builds because the `RUSTFLAGS` between tarpaulin and normal builds is different, per the article. The CI should use the same flag to make use of the cache efficienctly.

`--skip-clean` seams useless.

It seems you can configure a lot in `tarpaulin.toml` (like the target dir). I would like to have everything in this file, so that we have the same configuration locally.

Roms1383 replied:
> I understand now. So setting target-dir allows separate build artifacts between a normal build and a test build.

Yes indeed.

> It seems that the cache is not saved between those builds because the RUSTFLAGS between tarpaulin and normal builds is different, per the article. The CI should use the same flag to make use of the cache efficienctly.

The article says "one way or the other", but it could be worth trying.

> --skip-clean seams useless.

Agreed.

> It seems you can configure a lot in tarpaulin.toml (like the target dir). I would like to have everything in this file, so that we have the same configuration locally.

I won't be available for a couple of hours, but please go ahead and give it a try right if you want :)

Odonno replied:
Yeah. I re-read that article and read the other articles. Reducing build times seems to be a really complex topic.

I am aware that, for this kind of project, tests should run in 3 minutes and not 30 minutes. I think this is an infrastructure issue and that it can be improved in the future. In the meantime, we can rollback this changes and I can create an issue regarding tests build times.

What do you think?

Odonno replied:
Why not?

Roms1383 replied:
So you're suggesting to completely remove coverage job or? Not sure I understand.

Roms1383 replied:
If you want you can reach [me out on Discord](Roms1383#0211) to discuss it briefly too.

Roms1383 commented on .github/workflows/release.yml:
```diff
@@ -33,3 +39,15 @@ jobs:
+
+  publish:
+    needs: release
+    if: github.ref == 'refs/heads/main' && startsWith(github.ref, 'refs/tags/v')
```
you do not want to publish on each and every commit to each and every branch, do you?

Roms1383 commented on Cargo.toml:
```diff
@@ -15,26 +15,26 @@ edition = "2021"
 
 [dependencies]
-anyhow = "1.0.70"
+anyhow = "1"
```
yes, for most of the crates that's the case, but usually for "foundation" crates you can be confident that they are not gonna push breaking changes on a patch or a minor. I don't mind setting entire semver tho.

Roms1383 replied:
done in f4e1ec6

Roms1383 commented on Cargo.toml:
```diff
@@ -15,26 +15,26 @@ edition = "2021"
-sqlparser = "0.33.0"
-surrealdb = "1.0.0-beta.9"
+regex = "1"
+rust-ini = "0.19"
```
If there was any breaking change your tests would have caught it.

Roms1383 replied:
done in f4e1ec6

Roms1383 commented on samples/wasm/rsw-hello/src/lib.rs:
```diff
@@ -55,7 +55,7 @@ pub async fn get_blog_posts() -> js_sys::Array {
     posts
         .into_iter()
-        .map({ |post| JsValue::from_serde(&post).unwrap() })
+        .map(|post| serde_wasm_bindgen::to_value(&post).unwrap())
```
That's what `clippy` suggested, I just applied the suggestion. To check it out, simply revert the change locally and re-run `cargo clippy`.

Roms1383 replied:
Ah sorry, it's actually not `clippy` which yields the warning, but a build.
You can reproduce by running `(cd samples/wasm/rsw-hello && wasm-pack build --target web)` and here's what you get:

```sh
warning: use of deprecated associated function `wasm_bindgen::JsValue::from_serde`: causes dependency cycles, use `serde-wasm-bindgen` or `gloo_utils::format::JsValueSerdeExt` instead
```

Roms1383 commented on src/io.rs:
```diff
@@ -240,13 +230,10 @@ fn extract_string_dir_entry_value(
-            _ => None,
-        },
-        _ => None,
+    if let Some(DirEntryValue::String(value)) = f.get(&entry_attribute) {
```
Same, I just applied the `clippy` suggestion here. If there was any problem, your tests would have failed.

Roms1383 replied:
By the way

```rs
if let Some(DirEntryValue::Boolean(value)) = f.get(&entry_attribute) {
    return Some(value);
}
None
```

and

```rs
match f.get(&entry_attribute) {
    Some(DirEntryValue::Boolean(value)) => Some(value),
    _ => None,
}
```

are strictly equivalent.

Roms1383 replied:
standardised in 338b5db

Roms1383 commented on .github/workflows/main.yml:
```diff
@@ -1,57 +1,70 @@
+        run: cargo check --all
+
+      - name: Build WASM example
+        run: (cd samples/wasm/rsw-hello && wasm-pack build --target web)
```
done in 8d91ab8

Odonno commented on .github/workflows/example.yml:
```diff
@@ -0,0 +1,26 @@
+name: Example Workflow
+
+on: push
```
Add path filter on `samples/**`

https://stackoverflow.com/a/63822945

Roms1383 replied:
the problem if you do that, is whenever you'll make a change on `surrealdb-migrations` which breaks `samples`, then you won't see it until your next push to `samples` (which generally means 'never' since sample do not get changed very often).

Roms1383 replied:
what I did instead is to ignore some paths unrelated to `samples` in 811db37

Odonno replied:
Ah, yes, you're right.

Odonno commented on .github/workflows/example.yml:
```diff
@@ -0,0 +1,26 @@
+name: Example Workflow
```
Must be `Samples Workflow` to use the same name as the `samples` folder. Also rename the file name.

Roms1383 replied:
done in 4160075

Roms1383 replied:
and renamed in 1d24614

Odonno reviewed (changes requested):
I think we are close to the end. 👍

Odonno reviewed (approved):
Seems good to me. Thank you for your effort.
